File: Tests/scripts/gen2molcas6.py
#!/usr/bin/python

# Use 1 or 2 arguments:
#  1 (read):  generic input file
#  2 (write): Molcas input file (default: append ".mol" to generic input)

import sys

#=============================
# CHANGE THIS DATA AS NEEDED

Q_root = 1

seward_header = """\
*** Usa el directorio de bases (6-31G* con funciones esfericas)
!ln -sf $BASISDIR BASLIB
&SEWARD
  Title
    N-fenilpirrol
  >>> coord inline basis sph-31G*
  NoSymm"""

main_block = """\
!ln -sf $SaveDir/$Project.JobIph JOBOLD
&RASSCF
  JobIph
  NActElectrons = 12 0 0
  Inactive = 32
  Ras2 = 11
  CIRoot = 5 5 1
  RlxRoot = %(root)i
  OutOrbitals = Natural
    5
End of input
!rm -f JOBOLD

!cp $Project.rasscf.%(root)i.molden $Project.esp.molden
!cp $Project.RasOrb%(root_b)s $Project.esp.RasOrb

&RASSI
  NrOfJobIphs = 1  5
    1 2 3 4 5
End of input
""" % {"root": Q_root, "root_b": Q_root if (Q_root > 1) else "" }

alaska = """\
&MCLR
  SALa = %(root)i
  Iterations = 200
End of input

&ALASKA
End of input
""" % {"root": Q_root}

mckinley = """\
&MCKINLEY
End of input
!cat $Project.UnSym
"""

potential = """\
!cp $Project.JobIph $Project.JobOld

&RASSCF
  JobIph
  CIRestart
  CIOnly
  NActElectrons = 12 0 0
  Inactive = 32
  Ras2 = 11
  CIRoot = 1 %(root)i
    %(root)i
End of input

!cp $Project.JobOld $Project.JobIph
""" % {"root": Q_root}

#=============================
# Get input files

try:
  gen_input = sys.argv[1]
except IndexError:
  sys.exit("Missing input file")

try:
  mol_input = sys.argv[2]
except IndexError:
  mol_input = gen_input + ".mol"

#=============================
# Read the data from the generic input file

angstrom = 1.88972613289
Q_derivative = 0
Q_natoms = 0
Q_molecule = []
Q_chargesfile = ""
Q_charges = 0
Q_charges_points = []
Q_potential = 0
Q_potential_points = []

file_gen = open(gen_input, "r")
for line in file_gen:

  # Read the order of the energy derivative needed
  if (line.rstrip() == "Derivative"):
    Q_derivative = int(file_gen.next())

  # Read the geometry
  #  atom name, atomic symbol, atomic number, coordinates in Angstrom
  elif (line.rstrip() == "Geometry"):
    Q_natoms = int(file_gen.next())
    for i in range(Q_natoms):
      tmp = dict(zip(("name","type","number","x","y","z"),file_gen.next().split()))
      tmp["number"] = int(tmp["number"])
      tmp["x"] = float(tmp["x"])
      tmp["y"] = float(tmp["y"])
      tmp["z"] = float(tmp["z"])
      Q_molecule.append(tmp)

  # Read the filename containing external charges, in Angstrom
  elif (line.rstrip() == "External charge file"):
    Q_chargesfile = file_gen.next().rstrip()

  # ... or read the external charges positions and values, in Angstrom
  elif (line.rstrip() == "External charges"):
    Q_charges = int(file_gen.next())
    for i in range(Q_charges):
      tmp = dict(zip(("x","y","z","q"),file_gen.next().split()))
      tmp["x"] = float(tmp["x"])
      tmp["y"] = float(tmp["y"])
      tmp["z"] = float(tmp["z"])
      tmp["q"] = float(tmp["q"])
      Q_charges_points.append(tmp)

  # Read the points where the electrostatic potential is to be calculated, in Angstrom
  elif (line.rstrip() == "Potential points"):
    Q_potential = int(file_gen.next())
    for i in range(Q_potential):
      tmp = dict(zip(("x","y","z"),file_gen.next().split()))
      tmp["x"] = float(tmp["x"])
      tmp["y"] = float(tmp["y"])
      tmp["z"] = float(tmp["z"])
      Q_potential_points.append(tmp)

file_gen.close()

#=============================
# Write the Molcas input file

file_mol = open(mol_input, "w")

# Write the SEWARD input
print >> file_mol, seward_header

# Write the molecular geometry, in Angstrom
print >> file_mol, "    %i\n" % Q_natoms
for atom in Q_molecule:
  print >> file_mol, "%16s %20.12f %20.12f %20.12f" % (atom["type"], atom["x"], atom["y"], atom["z"])
print >> file_mol, "  End of input"

# Write the external charges, in Bohr, or the charge file
if (Q_charges):
  print >> file_mol, "  XField"
  print >> file_mol, "    %i 1" % Q_charges
  for charge in Q_charges_points:
    print >> file_mol, "%20.12f %20.12f %20.12f %20.12f %3.1f %3.1f %3.1f" % \
                       (charge["x"]*angstrom, charge["y"]*angstrom, charge["z"]*angstrom, charge["q"], 0, 0, 0)
elif (Q_chargesfile):
  print >> file_mol, "  XField"
  print >> file_mol, "@$MOLCAS_SUBMIT_PWD/%s" % Q_chargesfile

# Write the potential points, in Bohr
# (Skip, to avoid printing the potential for every root)
# (use molden instead)
#if (Q_potential):
#  print >> file_mol, "  EPot"
#  print >> file_mol, "    %i" % Q_potential
#  for point in Q_potential_points:
#    print >> file_mol, "%20.12f %20.12f %20.12f" % (point["x"]*angstrom, point["y"]*angstrom, point["z"]*angstrom)

# End of SEWARD input
print >> file_mol, "End of input\n"

print >> file_mol, main_block
if (Q_derivative >= 1): print >> file_mol, alaska
if (Q_derivative == 2): print >> file_mol, mckinley

# Work-around to avoid useless printing of the potential for every root...
#print >> file_mol, seward_header
#print >> file_mol, "    %i\n" % Q_natoms
#for atom in Q_molecule:
#  print >> file_mol, "%16s %20.12f %20.12f %20.12f" % (atom["type"], atom["x"], atom["y"], atom["z"])
#print >> file_mol, "  End of input"
#if (Q_charges):
#  print >> file_mol, "  XField"
#  print >> file_mol, "    %i 1" % Q_charges
#  for charge in Q_charges_points:
#    print >> file_mol, "%20.12f %20.12f %20.12f %20.12f %3.1f %3.1f %3.1f" % \
#                       (charge["x"]*angstrom, charge["y"]*angstrom, charge["z"]*angstrom, charge["q"], 0, 0, 0)
#elif (Q_chargesfile):
#  print >> file_mol, "  XField"
#  print >> file_mol, "@$MOLCAS_SUBMIT_PWD/%s" % Q_chargesfile
#if (Q_potential):
#  print >> file_mol, "  EPot"
#  print >> file_mol, "    %i" % Q_potential
#  for point in Q_potential_points:
#    print >> file_mol, "%20.12f %20.12f %20.12f" % (point["x"]*angstrom, point["y"]*angstrom, point["z"]*angstrom)
#print >> file_mol, "End of input\n"
#print >> file_mol, potential

file_mol.close()

File: Tests/scripts/gen2molcas6.pl
#!/usr/bin/perl -w

use strict;

# Use 1 or 2 arguments:
#  1 (read):  generic input file
#  2 (write): Molcas input file (default: append ".mol" to generic input)

my ($gen_input, $mol_input);
($gen_input, $mol_input) = @ARGV;

$mol_input = "$gen_input.mol" unless ($mol_input);

#===========================
# Read the data from the generic input file

my $Q_natoms;
my @Q_molecule;
my $Q_derivative;
my $Q_potential;
my @Q_potential_points;
my $Q_chargesfile;
my $Q_charges;
my @Q_charges_points;

open GEN_INPUT, $gen_input or die "No se puede leer el fichero $gen_input";

while (<GEN_INPUT>) {
  # Read the geometry
  #  atom name, atomic symbol, atomic number, coordinates in Angstrom
  if ($_ =~ /^Geometry$/) {
    $_ = <GEN_INPUT>;
    @_ = split;
    $Q_natoms = $_[0];
    for (my $i=0; $i<$Q_natoms; $i++) {
      $_ = <GEN_INPUT>;
      @_ = split;
      $Q_molecule[$i]{name} = $_[0];
      $Q_molecule[$i]{type} = $_[1];
      $Q_molecule[$i]{number} = $_[2];
      $Q_molecule[$i]{x} = $_[3];
      $Q_molecule[$i]{y} = $_[4];
      $Q_molecule[$i]{z} = $_[5];
    }
  }
  # Read the order of the energy derivative needed
  if ($_ =~ /^Derivative$/) {
    $_ = <GEN_INPUT>;
    @_ = split;
    $Q_derivative = $_[0];
  }
  # Read the external charges positions and values
  if ($_ =~ /^External charges$/) {
    $_ = <GEN_INPUT>;
    @_ = split;
    $Q_charges = $_[0];
    for (my $i=0; $i<$Q_charges; $i++) {
      $_ = <GEN_INPUT>;
      @_ = split;
      $Q_charges_points[$i]{x} = $_[0];
      $Q_charges_points[$i]{y} = $_[1];
      $Q_charges_points[$i]{z} = $_[2];
      $Q_charges_points[$i]{q} = $_[3];
    }
  }
  # ... or read the filename containing external charges
  if ($_ =~ /^External charge file$/) {
    $_ = <GEN_INPUT>;
    @_ = split;
    $Q_chargesfile = $_[0];
  }
  # Read the points where the electrostatic potential is to be calculated
  if ($_ =~ /^Potential points$/) {
    $_ = <GEN_INPUT>;
    @_ = split;
    $Q_potential = $_[0];
    for (my $i=0; $i<$Q_potential; $i++) {
      $_ = <GEN_INPUT>;
      @_ = split;
      $Q_potential_points[$i]{x} = $_[0];
      $Q_potential_points[$i]{y} = $_[1];
      $Q_potential_points[$i]{z} = $_[2];
    }
  }
}

#===========================
# Write the Molcas input file

my $angstrom = 1.88972613289;

open MOL_INPUT, ">$mol_input";

# Write the SEWARD INPUT
print MOL_INPUT <<"EOF";
&SEWARD &END
  Title
    Job Name
  >>> coord inline basis 6-31G
  NoSymm
EOF

# Write the molecular geometry, in Angstrom
print MOL_INPUT "    $Q_natoms\n\n";
for (my $i=0; $i<$Q_natoms; $i++) {
  printf MOL_INPUT "    %16s %20.12f %20.12f %20.12f\n",
    $Q_molecule[$i]{type}, $Q_molecule[$i]{x}, $Q_molecule[$i]{y}, $Q_molecule[$i]{z};
}
print MOL_INPUT "  End of input\n";

# Write the external charges, in Angstrom, or the charge file
if ($Q_charges) {
  print MOL_INPUT "  XField\n";
  print MOL_INPUT "    $Q_charges 1\n";
  for (my $i=0; $i<$Q_charges; $i++) {
    printf MOL_INPUT "%20.12f %20.12f %20.12f %20.12f %3.1f %3.1f %3.1f\n",
      $Q_charges_points[$i]{x}, $Q_charges_points[$i]{y}, $Q_charges_points[$i]{z}, $Q_charges_points[$i]{q}, 0, 0, 0;
  }
} elsif ($Q_chargesfile) {
  print MOL_INPUT "  XField\n";
  printf MOL_INPUT "\@\$MOLCAS_SUBMIT_PWD/%s\n", $Q_chargesfile;
}

# Write the potential points, in Bohr
if ($Q_potential) {
  print MOL_INPUT "  EPot\n";
  print MOL_INPUT "    $Q_potential\n";
  for (my $i=0; $i<$Q_potential; $i++) {
    printf MOL_INPUT "%20.12f %20.12f %20.12f\n",
      $Q_potential_points[$i]{x}*$angstrom, $Q_potential_points[$i]{y}*$angstrom, $Q_potential_points[$i]{z}*$angstrom;
  }
}
print MOL_INPUT "End of input\n";

# Write the SCF input
print MOL_INPUT <<"EOF";

&SCF &END
End of input
EOF

my $electrons = 0;
$electrons += $Q_molecule[$_]{number} foreach (0 .. $#Q_molecule);
my $orbitals = $electrons/2;
# Write the RASSCF input
# Change as needed
#	print MOL_INPUT <<"EOF";
#	
#	&RASSCF &END
#	  Inactive = $orbitals
#	  Ras2 = 0
#	  NActElectrons = 0 0 0
#	End of input
#	EOF

# Write the ALASKA/MCKINLEY input
if ($Q_derivative >= 1) {
  print MOL_INPUT <<"EOF";

&ALASKA &END
End of input
EOF
}
if ($Q_derivative >= 2) {
  print MOL_INPUT <<"EOF";

&MCKINLEY &END
End of input
!cat \$Project.UnSym
EOF
}

close(MOL_INPUT);

exit 0;
